concatMapIterable does not work for SYNC fusion and empty iterable

For specific pipelines where SYNC is used for concatMapIterable  operator it fails with empty iterable.

## Expected Behavior
concatMapIterable should work for empty iterators ending up as empty Flux for example.

## Actual Behavior
NoSuchElementException thrown when inner iterator is empty:

```
java.lang.AssertionError: expectation "expectComplete" failed (expected: onComplete(); actual: onError(java.util.NoSuchElementException))
	at reactor.test.MessageFormatter.assertionError(MessageFormatter.java:115)
	at reactor.test.MessageFormatter.failPrefix(MessageFormatter.java:104)
	at reactor.test.MessageFormatter.fail(MessageFormatter.java:73)
	at reactor.test.MessageFormatter.failOptional(MessageFormatter.java:88)
	at reactor.test.DefaultStepVerifierBuilder.lambda$expectComplete$4(DefaultStepVerifierBuilder.java:344)
	at reactor.test.DefaultStepVerifierBuilder$SignalEvent.test(DefaultStepVerifierBuilder.java:2289)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onSignal(DefaultStepVerifierBuilder.java:1529)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onExpectation(DefaultStepVerifierBuilder.java:1477)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onError(DefaultStepVerifierBuilder.java:1129)
	at reactor.core.publisher.FluxFlattenIterable$FlattenIterableSubscriber.drainSync(FluxFlattenIterable.java:636)
	at reactor.core.publisher.FluxFlattenIterable$FlattenIterableSubscriber.drain(FluxFlattenIterable.java:721)
	at reactor.core.publisher.FluxFlattenIterable$FlattenIterableSubscriber.request(FluxFlattenIterable.java:303)
	at reactor.test.DefaultStepVerifierBuilder$DefaultVerifySubscriber.onSubscribe(DefaultStepVerifierBuilder.java:1161)
	at reactor.core.publisher.FluxFlattenIterable$FlattenIterableSubscriber.onSubscribe(FluxFlattenIterable.java:222)
	at reactor.core.publisher.FluxPeekFuseable$PeekFuseableSubscriber.onSubscribe(FluxPeekFuseable.java:178)
	at reactor.core.publisher.FluxJust.subscribe(FluxJust.java:68)
	at reactor.core.publisher.Flux.subscribe(Flux.java:8660)
	at reactor.test.DefaultStepVerifierBuilder$DefaultStepVerifier.toVerifierAndSubscribe(DefaultStepVerifierBuilder.java:891)
	at reactor.test.DefaultStepVerifierBuilder$DefaultStepVerifier.verify(DefaultStepVerifierBuilder.java:831)
	at reactor.test.DefaultStepVerifierBuilder$DefaultStepVerifier.verify(DefaultStepVerifierBuilder.java:823)
	at com.ca.apm.common.ReactorTest.concatMapIterableTest(ReactorTest.java:70)
```
## Steps to Reproduce
This unit test fails for latest version (it was working for older releases 3.4.x):
```java
    @Test
    public void concatMapIterableTest() throws IOException {
        StepVerifier.create(
            Flux.just(1)
                .log() // Without .log it works as it uses another branch in source code
                .concatMapIterable(i -> {
                    Iterator<Integer> it = Collections.emptyIterator();
                    return () -> it;
                }))
        .expectComplete()
        .verify();
    }
```

## Possible Solution

This line should most likelly be same as in async drain:
https://github.com/reactor/reactor-core/blob/main/reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java#L591
```java
isEmpty = itFinite && sp.estimateSize() == 0;
```

Async drain is hasNext check that cporrectly identify my case is empty:
https://github.com/reactor/reactor-core/blob/main/reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java#L385
```java
isEmpty = itFinite ? sp.estimateSize() == 0 : !hasNext(sp);
```

## Your Environment

* Reactor version(s) used: 3.5.1
* JVM version (`java -version`): openjdk version "17.0.4.1"
* OS and version (eg `uname -a`): Linux, Windows

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java
@@ -588,7 +588,8 @@ final class FluxFlattenIterable<T, R> extends InternalFluxOperator<T, R> impleme
 							sp = iterable.spliterator();
 							itFinite = FluxIterable.checkFinite(sp);
 
-							isEmpty = itFinite && sp.estimateSize() == 0;
+
+							isEmpty = itFinite ? sp.estimateSize() == 0 : !hasNext(sp);
 						}
 						catch (Throwable exc) {
 							resetCurrent();

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java
@@ -588,7 +588,6 @@ final class FluxFlattenIterable<T, R> extends InternalFluxOperator<T, R> impleme
 							sp = iterable.spliterator();
 							itFinite = FluxIterable.checkFinite(sp);
 
-
 							isEmpty = itFinite ? sp.estimateSize() == 0 : !hasNext(sp);
 						}
 						catch (Throwable exc) {

--- a/reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java
+++ b/reactor-core/src/main/java/reactor/core/publisher/FluxFlattenIterable.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016-2022 VMware Inc. or its affiliates, All Rights Reserved.
+ * Copyright (c) 2016-2023 VMware Inc. or its affiliates, All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxFlattenIterableTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxFlattenIterableTest.java
@@ -707,6 +707,24 @@ public class FluxFlattenIterableTest extends FluxOperatorTest<String, String> {
 
 	@ParameterizedTestWithName
 	@MethodSource("reactor.core.publisher.FluxIterableTest#factory")
+	public void testFluxIterableSyncFusionEmptyCase(Function<Flux, Flux> fn) {
+		Iterable<String> iterable = mock(Iterable.class);
+		Mockito.when(iterable.spliterator())
+		       .thenReturn(mock(Spliterator.class));
+
+		StepVerifier.create(
+				            Flux.just(1, 2)
+				                .flatMapIterable(__ -> iterable)
+				                .as(fn)
+				                .next()
+		            )
+		            .expectSubscription()
+		            .expectComplete()
+		            .verify();
+	}
+
+	@ParameterizedTestWithName
+	@MethodSource("reactor.core.publisher.FluxIterableTest#factory")
 	public void testFluxIterableErrorHasNext(Function<Flux, Flux> fn) {
 		Iterable<String> iterable = mock(Iterable.class);
 		Spliterator mock = mock(Spliterator.class);

--- a/reactor-core/src/test/java/reactor/core/publisher/FluxFlattenIterableTest.java
+++ b/reactor-core/src/test/java/reactor/core/publisher/FluxFlattenIterableTest.java
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2016-2022 VMware Inc. or its affiliates, All Rights Reserved.
+ * Copyright (c) 2016-2023 VMware Inc. or its affiliates, All Rights Reserved.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.

